README.md
<div align="center">
    <h1><a href="https://mongoaud.it"><img src="https://raw.githubusercontent.com/stampery/mongoaudit/master/rsc/github-header.png" alt="mongoaudit"/></a></h1>
    <a href="https://travis-ci.org/stampery/mongoaudit"><img src="https://travis-ci.org/stampery/mongoaudit.svg?branch=master"></a>
    <a href="https://landscape.io/github/stampery/mongoaudit/master"><img alt="Code Health" src="https://landscape.io/github/stampery/mongoaudit/master/landscape.svg?style=flat"/></a>
    <a href="https://codeclimate.com/repos/588f61f717e4fe24b80046f6/feed"><img alt="Code Climate" src="https://codeclimate.com/repos/588f61f717e4fe24b80046f6/badges/ed691ca1655c0eb8a4a5/gpa.svg" /></a>
    <a href="https://codeclimate.com/repos/588f61f717e4fe24b80046f6/feed"><img alt="Issue Count" src="https://codeclimate.com/repos/588f61f717e4fe24b80046f6/badges/ed691ca1655c0eb8a4a5/issue_count.svg" /></a>
    <br/><br/>
    <p><strong>mongoaudit</strong> is a CLI tool for auditing MongoDB servers, detecting poor security settings and performing automated penetration testing.</p>
</div>

<h2 align="center">Installing</h2>

Clone this repository and run the setup:

```bash
> git clone https://github.com/stampery/mongoaudit.git
> cd mongoaudit
> python setup.py install
> mongoaudit
```

<h2 align="center">Introduction</h2>
<p>It is widely known that there are quite a few holes in MongoDB's default configuration settings. This fact, combined with abundant lazy system administrators and developers, has led to what the press has called the <a href="http://thenextweb.com/insider/2017/01/08/mongodb-ransomware-exists-people-bad-security/"><i>MongoDB apocalypse</i></a>.</p>

<p><strong>mongoaudit</strong> not only detects misconfigurations, known vulnerabilities and bugs but also gives you advice on how to fix them, recommends best practices and teaches you how to DevOp like a pro! </p>

<p>This is how the actual app looks like:</p>

<p align="center">
    <img align="center" src="https://raw.githubusercontent.com/stampery/mongoaudit/master/rsc/screenshot.png" alt="mongoaudit screenshot"/>
    <br /><i>Yep, that's material design on a console line interface. (Powered by <a href="https://github.com/urwid/urwid">urwid</a>)</i>
</p>

<h2 align="center">Supported tests</h2>

* MongoDB listens on a port different to default one
* Server only accepts connections from whitelisted hosts / networks
* MongoDB HTTP status interface is not accessible on port 28017
* MongoDB is not exposing its version number
* MongoDB version is newer than 2.4
* TLS/SSL encryption is enabled
* Authentication is enabled
* SCRAM-SHA-1 authentication method is enabled
* Server-side Javascript is forbidden *
* Roles granted to the user only permit CRUD operations *
* The user has permissions over a single database *
* Security bug [CVE-2015-7882](https://jira.mongodb.org/browse/SERVER-20691)
* Security bug [CVE-2015-2705](https://jira.mongodb.org/browse/SERVER-17521)
* Security bug [CVE-2014-8964](https://jira.mongodb.org/browse/SERVER-17252)
* Security bug [CVE-2015-1609](https://jira.mongodb.org/browse/SERVER-17264)
* Security bug [CVE-2014-3971](https://jira.mongodb.org/browse/SERVER-13753)
* Security bug [CVE-2014-2917](https://jira.mongodb.org/browse/SERVER-13644)
* Security bug [CVE-2013-4650](https://jira.mongodb.org/browse/SERVER-9983)
* Security bug [CVE-2013-3969](https://jira.mongodb.org/browse/SERVER-9878)
* Security bug [CVE-2012-6619](https://www.cvedetails.com/cve/CVE-2012-6619/)
* Security bug [CVE-2013-1892](https://www.cvedetails.com/cve/CVE-2013-1892/)
* Security bug [CVE-2013-2132](https://www.cvedetails.com/cve/CVE-2013-2132/)


_Tests marked with an asterisk (`*`) require valid authentication credentials._


<h2 align="center">How can I best secure my MongoDB?</h2>

Once you run any of the test suites provided by __mongoaudit__, it will offer you to receive a fully detailed report via email. This personalized report links to a series of useful guides on how to fix every specific issue and how to harden your MongoDB deployments.

For your convenience, we have also published the __mongoaudit__ guides in our [Medium publication](https://medium.com/mongoaudit).

<h2 align="center">Contributing</h2>

We're happy you want to contribute! You can help us in different ways:

* Open an issue with suggestions for improvements and errors you're facing.
* Fork this repository and submit a pull request.
* Improve the documentation.

To submit a pull request, fork the mongoaudit repository and then clone your fork:

```bash
git clone git@github.com:<your-name>/mongoaudit.git
```

Make your suggested changes, `git push` and then submit a pull request.

<h2 align="center">Legal</h2>

<h3>License</h3>
<strong>mongoaudit</strong> is released under the [MIT License](https://github.com/stampery/mongoaudit/blob/master/LICENSE).

<h3>Disclaimer</h3>

    "With great power comes great responsibility"

* Never use this tool on servers you don't own. Unauthorized access to strangers' computer systems is a crime in many countries.
* Please use this tool is at your own risk. We will accept no liability for any loss or damage which you may incur no matter how caused.
* Don't be evil! :trollface:

<h3>Suport and trademarks</h3>
<i>This software is not supported or endorsed in any way by MongoDB Inc., Compose Inc., ObjectsLab Corporation nor other products or services providers it interoperates with. It neither tries to mimic or replace any software originally conceived by the owners of those products and services. In the same manner, any third party's trademark or intellectual property that may appear in this software must be understood as a strictly illustrative reference to the service provider it represents, and is never used in any way that may lead to confusion, so no abuse is intended.</i>

[<img style="width:100%;" src="https://raw.githubusercontent.com/trailbot/vault/master/dist/img/footer.png">](https://stampery.com)


mongoaudit/CHANGELOG
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.1.1] - 2020-12-03
### Fixed
- Bug fixing

## [0.1.0] - 2018-03-25
### Added
- Added Python 3 support

src/testers.py
# -*- coding: utf-8 -*-
import socket
import time

import pymongo

from src.tools import decode_to_string, in_range


class Tester(object):
    """
    Tester
    Args:
      cred (dict(str: str)): credentials
      tests (Test[]): tests to run
    """

    def __init__(self, cred, tests):
        self.cred = cred
        self.tests = [Test(t, self) for t in tests]
        self.conn = self.get_connection()
        self.info = self.get_info()
        self.database = None

    def run(self, each, end):
        """
        Args:
          each (): function to call before each test runs (used to update and redraw the screen)
          end (function): function to call when all the tests have finished running
        """
        result = []
        for test in self.tests:
            each(test)
            res = test.run()
            result.append(res)
            time.sleep(0.5)
            if test.breaks == bool(res['result']):
                break

        if len(result) < len(self.tests):
            result += [{
                'name': x.name,
                'severity': x.severity,
                'title': x.title,
                'caption': x.caption,
                'message': 'This test was omitted',
                'extra_data': None,
                'result': 3} for x in self.tests[len(result): len(self.tests)]]

        self.conn.close()
        end(result)

    def get_connection(self):
        """
        Returns:
          pymongo.MongoClient: client that takes as arguments (fqdn, port)

        """
        fqdn, port = self.cred['nodelist'][0]
        return pymongo.MongoClient(fqdn, port, serverSelectionTimeoutMS=1)

    def get_info(self):
        """
        Returns:
          dict() or None: Get information about the MongoDB server we’re connected to.
        Note:
          https://docs.mongodb.com/v3.2/reference/command/buildInfo/#dbcmd.buildInfo
        """
        if hasattr(self, 'info'):
            return self.info
        try:
            info = self.conn.server_info()
            self.info = info
            return self.info
        except pymongo.errors.ConnectionFailure:
            return None

    def get_db(self):
        """
        Authenticates to database
        Returns:
          pymongo.database.Database or False:  database(singleton) or false if authentication fails
        """
        if bool(self.database):
            return self.database
        try:
            database = self.conn[self.cred['database']]
            database.authenticate(self.cred['username'], self.cred['password'])
            self.database = database
            return self.database
        except (pymongo.errors.OperationFailure, ValueError, TypeError):
            return False

    def get_roles(self):
        data_base = self.get_db()
        return data_base.command('usersInfo', self.cred['username'])['users'][0]


class Test(object):
    def __init__(self, test, tester):
        """
        Args:
          test dict():
          tester Tester:

        Notes:
          message should contain an array with the messages for the different results that
          the test can return, if message is a function it should return a str[]
        """
        self.name, self.severity, self.title, self.caption, self.message = \
            test['test_name'], test['severity'], test['title'], test['caption'], test['messages']
        self.breaks = test['breaks'] if 'breaks' in test else None
        self.tester = tester

    def run(self):
        """
        Returns:
          tuple(int, str): test result value and message
        """
        result = TEST_FUNCTIONS[self.name](self)
        message = self.message

        if isinstance(result, list):
            value = result[0]
            extra_data = result[1]
        else:
            value = result
            extra_data = None

        return {'name': self.name, 'severity': self.severity, 'title': self.title,
                'caption': self.caption, 'message': message[value], 'result': value,
                'extra_data': extra_data}


def try_socket(test, forced_port=None):
    fqdn, port = test.tester.cred['nodelist'][0]
    try:
        socket.create_connection((fqdn, forced_port or port), 5)
    except socket.error:
        return True
    else:
        return False


def try_authorization(test):
    try:
        test.tester.conn.database_names()
    except pymongo.errors.OperationFailure:
        return True
    else:
        return False


def try_credentials(test):
    """
    verify if the credentials are valid
    """
    return bool(test.tester.get_db())


def try_javascript(test):
    """
    Check if javascript is enabled in MongoDB
    """
    try:
        database = test.tester.get_db()
        database.test.find_one({"$where": "function() { return true;}"})
    except pymongo.errors.OperationFailure:
        return True
    else:
        return False


def try_roles(test):
    """
    Verify that the user roles are not administrative
    Returns:
      [bool or 2 , str ]: True Valid role, False invalid role, 2 custome role ,
    """

    def valid_role(role):
        """
        Args:
          role (str): name of a role
        Returns:
          Bool: True if the role is not administrative
        """
        return role not in [
            'userAdminAnyDatabase',
            'dbAdminAnyDatabase'
            'dbAdmin',
            'dbOwner',
            'userAdmin',
            'clusterAdmin']

    def result_default_value():
        """
        Returns:
          dict(set()): returns an empty dictionary that contains 3 empty
          sets where valid, invalid and custom roles are stored
        """
        return {'invalid': set([]), 'valid': set([]), 'custom': set([])}

    def combine_result(value_1, value_2):
        """
        Args:
          value_1 (dict(set())): result_default_value
          value_2 (dict(set())): result_default_value
        Returns:
          dict(set()): the union of 2 default values
        """
        return {'invalid': value_1['invalid'].union(value_2['invalid']),
                'valid': value_1['valid'].union(value_2['valid']),
                'custom': value_1['custom'].union(value_2['custom'])}

    def basic_validation(roles):
        """
        Basic validation in case validate role fails due to lack of permissions
        Returns:
          result_default_value
        """
        validated = result_default_value()
        for role in roles['roles']:
            if valid_role(role['role']):
                validated['valid'].add(role['role'])
            else:
                validated['invalid'].add(role['role'])
        return validated

    def get_roles(role, key):
        return validate_role(role[key]) if key in role else result_default_value()

    def get_builtin_role(role):
        result = result_default_value()
        is_valid_role = valid_role(role['role'])
        if is_valid_role and role['isBuiltin']:
            result['valid'].add(role['role'])
        elif is_valid_role:
            result['custom'].add(role['role'])
        else:
            result['invalid'].add(role['role'])
        inherited = get_roles(role, 'inherited')
        other_roles = get_roles(role, 'roles')
        return combine_result(result, combine_result(inherited, other_roles))

    def validate_role_dict(role):
        if 'role' in role:
            return get_builtin_role(role) if 'isBuiltin' in role else  \
                validate_role(database.command('rolesInfo', role)['roles'])
        if 'roles' in role and bool(role['roles']):
            return validate_role(role['roles'])
        else:
            raise Exception('Non exhaustive type case')

    def validate_role(role):
        """
        Recursively process the different roles that a role implements or inherits
        Args:
          role (list or dict): value returned by database.command 'usersInfo' or 'rolesInfo'
        """
        if isinstance(role, list):
            return reduce(lambda x, y: combine_result(x, y), [validate_role(r) for r in role])\
                if bool(role) else result_default_value()
        elif isinstance(role, dict):
            return validate_role_dict(role)
        else:
            raise Exception('Non exhaustive type case')

    def get_message(state, text1, text2):
        return text1 + decode_to_string(validated[state]) + text2

    def validation_result(validated):
        if bool(validated['invalid']):
            # if the profile is invalid
            return [False, decode_to_string(validated['invalid'])]
        elif bool(validated['custom']):
            # if the profile has custom permissions
            message = get_message('valid', 'Your user\'s role set ',
                                  ' seem to be ok, but we couldn\'t do an exhaustive check.')
            return [2, message]
        # if everything seems to be ok
        return [True, decode_to_string(validated['valid'])]



    database = test.tester.get_db()
    roles = test.tester.get_roles()
    validated = {}

    try:
        validated = validate_role(roles)
    except pymongo.errors.OperationFailure:
        # if the users doesn't have permission to run the command 'rolesInfo'
        validated = basic_validation(roles)
        if bool(validated['valid']):
            message = get_message('valid', 'You user permission ',
                                  ' didn\'t allow us to do an exhaustive check')
            return [2, message]

    return validation_result(validated)





def try_dedicated_user(test):
    """
    Verify that the role only applies to one database
    """
    roles = test.tester.get_roles()
    user_role_dbs = set()
    for role in roles['roles']:
        user_role_dbs.add(role['db'])

    return [bool(len(user_role_dbs)), decode_to_string(user_role_dbs)]


def try_scram(test):
    try:
        conn = test.tester.get_connection()
        cred = test.tester.cred
        return conn[cred["database"]]\
            .authenticate(cred["username"], cred["password"], mechanism='SCRAM-SHA-1')
    except (pymongo.errors.OperationFailure, ValueError, TypeError):
        return False


# the following functions are for https://www.mongodb.com/alerts security related


def alerts_dec012015(test):
    if "modules" in test.tester.info:
        enterprise = "enterprise" in test.tester.info["modules"]
        version = test.tester.info["version"]
        return not (enterprise and in_range(version, "3.0.0", "3.0.6"))
    else:
        return True


def alerts_mar272015(test):
    return test.tester.info["version"] != "3.0.0"


def alerts_mar252015(test):
    version = test.tester.info["version"]
    return version > "2.6.8" and version != "3.0.0"


def alerts_feb252015(test):
    version = test.tester.info["version"]
    return version > "2.6.7" or version == "2.4.13"


def alerts_jun172014(test):
    version = test.tester.info["version"]
    return version not in ["2.6.0", "2.6.1"]


def alerts_may052014(test):
    version = test.tester.info["version"]
    return version != "2.6.0"


def alerts_jun202013(test):
    version = test.tester.info["version"]
    return not in_range(version, "2.4.0", "2.4.4") and version != "2.5.1"


def alerts_jun052013(test):
    version = test.tester.info["version"]
    return not in_range(version, "2.4.0", "2.4.4")


def alerts_mar062014(test):
    version = test.tester.info["version"]
    return version > "2.3.1"


def alerts_oct012013(test):
    version = test.tester.info["version"]
    return version > "2.2.3"


def alerts_aug152013(test):
    version = test.tester.info["version"]
    return version > "2.5.1"


TEST_FUNCTIONS = {
    "1": lambda test: not (test.tester.cred['nodelist'][0][1] == 27017 and bool(test.tester.info)),
    "2": try_socket,
    "3": lambda test: try_socket(test, 28017),
    "4": lambda test: "version" not in test.tester.info,
    "5": lambda test: [test.tester.info["version"] > "2.4", str(test.tester.info["version"])],
    "6": lambda test: bool(test.tester.info["OpenSSLVersion"])
         if ("OpenSSLVersion" in test.tester.info)
         else test.tester.info["openssl"]["running"] != "disabled",
    "7": try_authorization,
    "8": lambda test: bool(test.tester.get_db()),
    "9": try_javascript,
    "10": try_roles,
    "11": try_dedicated_user,
    "12": try_scram,
    "13": alerts_dec012015,
    "14": alerts_mar272015,
    "15": alerts_mar252015,
    "16": alerts_feb252015,
    "17": alerts_jun172014,
    "18": alerts_may052014,
    "19": alerts_jun202013,
    "20": alerts_jun052013,
    "21": alerts_mar062014,
    "22": alerts_oct012013,
    "23": alerts_aug152013
}
